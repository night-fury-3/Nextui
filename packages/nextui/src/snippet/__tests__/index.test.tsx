import React from 'react';
import { mount } from 'enzyme';
import Snippet from '../index';

const command = 'yarn add @nextui-org/react';
const multiLine = ['cd project', 'now'];

describe('Snippet', () => {
  beforeAll(() => {
    window.getSelection = jest.fn().mockImplementation(() => ({
      removeAllRanges: jest.fn(),
      addRange: jest.fn()
    }));
    document.createRange = jest.fn().mockImplementation(() => ({
      selectNode: jest.fn()
    }));
  });

  it('should render correctly', () => {
    const wrapper = mount(<Snippet text={command} />);
    expect(wrapper.html()).toMatchSnapshot();
    expect(() => wrapper.unmount()).not.toThrow();
  });

  it('should work with multi-line', () => {
    const wrapper = mount(<Snippet text={multiLine} />);
    expect(wrapper.html()).toMatchSnapshot();
    expect(() => wrapper.unmount()).not.toThrow();
  });

  it('should work with different styles', () => {
    const wrapper = mount(
      <div>
        <Snippet text={command} />
        <Snippet text={command} filled />
        <Snippet text={command} color="secondary" />
        <Snippet text={command} color="success" filled />
      </div>
    );
    expect(wrapper.html()).toMatchSnapshot();
    expect(() => wrapper.unmount()).not.toThrow();
  });

  it('text should be copied', () => {
    document.execCommand = jest.fn();
    const wrapper = mount(<Snippet text={command} />);
    wrapper.find('.nextui-snippet-copy-button').at(0).simulate('click');
    expect(document.execCommand).toHaveBeenCalled();
    (document.execCommand as jest.Mock).mockRestore();
  });

  it('multi-line commands should be copied', () => {
    document.execCommand = jest.fn();
    const wrapper = mount(<Snippet text={multiLine} />);
    wrapper.find('.nextui-snippet-copy-button').at(0).simulate('click');
    expect(document.execCommand).toHaveBeenCalled();
    (document.execCommand as jest.Mock).mockRestore();
  });

  it('child commands should be copied', () => {
    document.execCommand = jest.fn();
    const wrapper = mount(<Snippet>{command}</Snippet>);
    wrapper.find('.nextui-snippet-copy-button').at(0).simulate('click');
    expect(document.execCommand).toHaveBeenCalled();
    (document.execCommand as jest.Mock).mockRestore();
  });

  it('should disable copy function', () => {
    const wrapper = mount(<Snippet text={command} copy="prevent" />);
    expect(wrapper.find('.nextui-snippet-copy-button').length).toBe(0);
  });

  it('should work with custom symbol', () => {
    const wrapper = mount(<Snippet text={command} symbol=">" />);
    expect(wrapper.html()).toMatchSnapshot();
    expect(() => wrapper.unmount()).not.toThrow();

    const emptySymbolWrapper = mount(<Snippet text={command} symbol=" " />);
    expect(emptySymbolWrapper.html()).toMatchSnapshot();

    const emptySymbolWrapper2 = mount(<Snippet text={command} symbol="" />);
    expect(emptySymbolWrapper2.html()).toEqual(emptySymbolWrapper.html());

    expect(() => emptySymbolWrapper.unmount()).not.toThrow();
    expect(() => emptySymbolWrapper2.unmount()).not.toThrow();
  });

  it('should work with custom toast', () => {
    document.execCommand = jest.fn();
    const wrapper = mount(
      <Snippet text={command} tooltipCopiedText="Code copied!" />
    );
    wrapper.find('.nextui-snippet-copy-button').at(0).simulate('click');
    expect(document.execCommand).toHaveBeenCalled();
    (document.execCommand as jest.Mock).mockRestore();
  });

  afterAll(() => {
    (window.getSelection as jest.Mock).mockRestore();
    (document.createRange as jest.Mock).mockRestore();
  });
});
