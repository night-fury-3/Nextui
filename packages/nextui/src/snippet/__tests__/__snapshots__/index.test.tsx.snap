// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snippet should render correctly 1`] = `
"<div class=\\"snippet \\"><pre>yarn add @nextui-org/react</pre><div role=\\"button\\" tabindex=\\"-1\\" class=\\"tooltip \\"><span class=\\"copy\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"18\\" height=\\"18\\" viewBox=\\"0 0 24 24\\"><path fill=\\"#444\\" d=\\"M20 2H10c-1.103 0-2 .897-2 2v4H4c-1.103 0-2 .897-2 2v10c0 1.103.897 2 2 2h10c1.103 0 2-.897 2-2v-4h4c1.103 0 2-.897 2-2V4c0-1.103-.897-2-2-2zM4 20V10h10l.002 10H4zm16-6h-4v-4c0-1.103-.897-2-2-2h-4V4h10v10z\\"></path></svg></span><style>
        .tooltip {
          width: max-content;
          display: inherit;
        }
      </style></div><style>
        .snippet {
          display: flex;
          position: relative;
          width: initial;
          max-width: 100%;
          padding: calc(20px * 0.75) 20px;
          color: #000;
          background: #fff;
          border: 0px solid #eaeaea;
          border-radius: 14px;
        }
        pre {
          margin: 0;
          padding: 0;
          border: none;
          border-radius: 0;
          width: calc(100% - 2 * 20px);
          background-color: transparent;
          color: #000;
          font-size: 0.8125rem;
        }
        pre::before {
          content: '$ ';
          user-select: none;
        }
        pre :global(*) {
          margin: 0;
          padding: 0;
          font-size: inherit;
          color: inherit;
        }
        .copy {
          background-color: #fff;
          display: inline-flex;
          justify-content: center;
          align-items: flex-start;
          width: calc(2 * 20px);
          color: inherit;
          transition: opacity 0.2s ease 0s;
          cursor: pointer;
          user-select: none;
        }
        .copy:hover {
          opacity: 0.7;
        }
      </style></div>"
`;

exports[`Snippet should work with custom symbol 1`] = `
"<div class=\\"snippet \\"><pre>yarn add @nextui-org/react</pre><div role=\\"button\\" tabindex=\\"-1\\" class=\\"tooltip \\"><span class=\\"copy\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"18\\" height=\\"18\\" viewBox=\\"0 0 24 24\\"><path fill=\\"#444\\" d=\\"M20 2H10c-1.103 0-2 .897-2 2v4H4c-1.103 0-2 .897-2 2v10c0 1.103.897 2 2 2h10c1.103 0 2-.897 2-2v-4h4c1.103 0 2-.897 2-2V4c0-1.103-.897-2-2-2zM4 20V10h10l.002 10H4zm16-6h-4v-4c0-1.103-.897-2-2-2h-4V4h10v10z\\"></path></svg></span><style>
        .tooltip {
          width: max-content;
          display: inherit;
        }
      </style></div><style>
        .snippet {
          display: flex;
          position: relative;
          width: initial;
          max-width: 100%;
          padding: calc(20px * 0.75) 20px;
          color: #000;
          background: #fff;
          border: 0px solid #eaeaea;
          border-radius: 14px;
        }
        pre {
          margin: 0;
          padding: 0;
          border: none;
          border-radius: 0;
          width: calc(100% - 2 * 20px);
          background-color: transparent;
          color: #000;
          font-size: 0.8125rem;
        }
        pre::before {
          content: '> ';
          user-select: none;
        }
        pre :global(*) {
          margin: 0;
          padding: 0;
          font-size: inherit;
          color: inherit;
        }
        .copy {
          background-color: #fff;
          display: inline-flex;
          justify-content: center;
          align-items: flex-start;
          width: calc(2 * 20px);
          color: inherit;
          transition: opacity 0.2s ease 0s;
          cursor: pointer;
          user-select: none;
        }
        .copy:hover {
          opacity: 0.7;
        }
      </style></div>"
`;

exports[`Snippet should work with custom symbol 2`] = `
"<div class=\\"snippet \\"><pre>yarn add @nextui-org/react</pre><div role=\\"button\\" tabindex=\\"-1\\" class=\\"tooltip \\"><span class=\\"copy\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"18\\" height=\\"18\\" viewBox=\\"0 0 24 24\\"><path fill=\\"#444\\" d=\\"M20 2H10c-1.103 0-2 .897-2 2v4H4c-1.103 0-2 .897-2 2v10c0 1.103.897 2 2 2h10c1.103 0 2-.897 2-2v-4h4c1.103 0 2-.897 2-2V4c0-1.103-.897-2-2-2zM4 20V10h10l.002 10H4zm16-6h-4v-4c0-1.103-.897-2-2-2h-4V4h10v10z\\"></path></svg></span><style>
        .tooltip {
          width: max-content;
          display: inherit;
        }
      </style></div><style>
        .snippet {
          display: flex;
          position: relative;
          width: initial;
          max-width: 100%;
          padding: calc(20px * 0.75) 20px;
          color: #000;
          background: #fff;
          border: 0px solid #eaeaea;
          border-radius: 14px;
        }
        pre {
          margin: 0;
          padding: 0;
          border: none;
          border-radius: 0;
          width: calc(100% - 2 * 20px);
          background-color: transparent;
          color: #000;
          font-size: 0.8125rem;
        }
        pre::before {
          content: '';
          user-select: none;
        }
        pre :global(*) {
          margin: 0;
          padding: 0;
          font-size: inherit;
          color: inherit;
        }
        .copy {
          background-color: #fff;
          display: inline-flex;
          justify-content: center;
          align-items: flex-start;
          width: calc(2 * 20px);
          color: inherit;
          transition: opacity 0.2s ease 0s;
          cursor: pointer;
          user-select: none;
        }
        .copy:hover {
          opacity: 0.7;
        }
      </style></div>"
`;

exports[`Snippet should work with different styles 1`] = `
"<div><div class=\\"snippet \\"><pre>yarn add @nextui-org/react</pre><div role=\\"button\\" tabindex=\\"-1\\" class=\\"tooltip \\"><span class=\\"copy\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"18\\" height=\\"18\\" viewBox=\\"0 0 24 24\\"><path fill=\\"#444\\" d=\\"M20 2H10c-1.103 0-2 .897-2 2v4H4c-1.103 0-2 .897-2 2v10c0 1.103.897 2 2 2h10c1.103 0 2-.897 2-2v-4h4c1.103 0 2-.897 2-2V4c0-1.103-.897-2-2-2zM4 20V10h10l.002 10H4zm16-6h-4v-4c0-1.103-.897-2-2-2h-4V4h10v10z\\"></path></svg></span><style>
        .tooltip {
          width: max-content;
          display: inherit;
        }
      </style></div><style>
        .snippet {
          display: flex;
          position: relative;
          width: initial;
          max-width: 100%;
          padding: calc(20px * 0.75) 20px;
          color: #000;
          background: #fff;
          border: 0px solid #eaeaea;
          border-radius: 14px;
        }
        pre {
          margin: 0;
          padding: 0;
          border: none;
          border-radius: 0;
          width: calc(100% - 2 * 20px);
          background-color: transparent;
          color: #000;
          font-size: 0.8125rem;
        }
        pre::before {
          content: '$ ';
          user-select: none;
        }
        pre :global(*) {
          margin: 0;
          padding: 0;
          font-size: inherit;
          color: inherit;
        }
        .copy {
          background-color: #fff;
          display: inline-flex;
          justify-content: center;
          align-items: flex-start;
          width: calc(2 * 20px);
          color: inherit;
          transition: opacity 0.2s ease 0s;
          cursor: pointer;
          user-select: none;
        }
        .copy:hover {
          opacity: 0.7;
        }
      </style></div><div class=\\"snippet \\"><pre>yarn add @nextui-org/react</pre><div role=\\"button\\" tabindex=\\"-1\\" class=\\"tooltip \\"><span class=\\"copy\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"18\\" height=\\"18\\" viewBox=\\"0 0 24 24\\"><path fill=\\"#444\\" d=\\"M20 2H10c-1.103 0-2 .897-2 2v4H4c-1.103 0-2 .897-2 2v10c0 1.103.897 2 2 2h10c1.103 0 2-.897 2-2v-4h4c1.103 0 2-.897 2-2V4c0-1.103-.897-2-2-2zM4 20V10h10l.002 10H4zm16-6h-4v-4c0-1.103-.897-2-2-2h-4V4h10v10z\\"></path></svg></span><style>
        .tooltip {
          width: max-content;
          display: inherit;
        }
      </style></div><style>
        .snippet {
          display: flex;
          position: relative;
          width: 20%;
          max-width: 100%;
          padding: calc(20px * 0.75) 20px;
          color: #000;
          background: #fff;
          border: 0px solid #eaeaea;
          border-radius: 14px;
        }
        pre {
          margin: 0;
          padding: 0;
          border: none;
          border-radius: 0;
          width: calc(100% - 2 * 20px);
          background-color: transparent;
          color: #000;
          font-size: 0.8125rem;
        }
        pre::before {
          content: '$ ';
          user-select: none;
        }
        pre :global(*) {
          margin: 0;
          padding: 0;
          font-size: inherit;
          color: inherit;
        }
        .copy {
          background-color: #fff;
          display: inline-flex;
          justify-content: center;
          align-items: flex-start;
          width: calc(2 * 20px);
          color: inherit;
          transition: opacity 0.2s ease 0s;
          cursor: pointer;
          user-select: none;
        }
        .copy:hover {
          opacity: 0.7;
        }
      </style></div><div class=\\"snippet \\"><pre>yarn add @nextui-org/react</pre><div role=\\"button\\" tabindex=\\"-1\\" class=\\"tooltip \\"><span class=\\"copy\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"18\\" height=\\"18\\" viewBox=\\"0 0 24 24\\"><path fill=\\"#444\\" d=\\"M20 2H10c-1.103 0-2 .897-2 2v4H4c-1.103 0-2 .897-2 2v10c0 1.103.897 2 2 2h10c1.103 0 2-.897 2-2v-4h4c1.103 0 2-.897 2-2V4c0-1.103-.897-2-2-2zM4 20V10h10l.002 10H4zm16-6h-4v-4c0-1.103-.897-2-2-2h-4V4h10v10z\\"></path></svg></span><style>
        .tooltip {
          width: max-content;
          display: inherit;
        }
      </style></div><style>
        .snippet {
          display: flex;
          position: relative;
          width: initial;
          max-width: 100%;
          padding: calc(20px * 0.75) 20px;
          color: #7928ca;
          background: #fff;
          border: 0px solid #7928ca;
          border-radius: 14px;
        }
        pre {
          margin: 0;
          padding: 0;
          border: none;
          border-radius: 0;
          width: calc(100% - 2 * 20px);
          background-color: transparent;
          color: #7928ca;
          font-size: 0.8125rem;
        }
        pre::before {
          content: '$ ';
          user-select: none;
        }
        pre :global(*) {
          margin: 0;
          padding: 0;
          font-size: inherit;
          color: inherit;
        }
        .copy {
          background-color: #fff;
          display: inline-flex;
          justify-content: center;
          align-items: flex-start;
          width: calc(2 * 20px);
          color: inherit;
          transition: opacity 0.2s ease 0s;
          cursor: pointer;
          user-select: none;
        }
        .copy:hover {
          opacity: 0.7;
        }
      </style></div><div class=\\"snippet \\"><pre>yarn add @nextui-org/react</pre><div role=\\"button\\" tabindex=\\"-1\\" class=\\"tooltip \\"><span class=\\"copy\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"18\\" height=\\"18\\" viewBox=\\"0 0 24 24\\"><path fill=\\"#444\\" d=\\"M20 2H10c-1.103 0-2 .897-2 2v4H4c-1.103 0-2 .897-2 2v10c0 1.103.897 2 2 2h10c1.103 0 2-.897 2-2v-4h4c1.103 0 2-.897 2-2V4c0-1.103-.897-2-2-2zM4 20V10h10l.002 10H4zm16-6h-4v-4c0-1.103-.897-2-2-2h-4V4h10v10z\\"></path></svg></span><style>
        .tooltip {
          width: max-content;
          display: inherit;
        }
      </style></div><style>
        .snippet {
          display: flex;
          position: relative;
          width: initial;
          max-width: 100%;
          padding: calc(20px * 0.75) 20px;
          color: #fff;
          background: #17c964;
          border: 0px solid #17c964;
          border-radius: 14px;
        }
        pre {
          margin: 0;
          padding: 0;
          border: none;
          border-radius: 0;
          width: calc(100% - 2 * 20px);
          background-color: transparent;
          color: #fff;
          font-size: 0.8125rem;
        }
        pre::before {
          content: '$ ';
          user-select: none;
        }
        pre :global(*) {
          margin: 0;
          padding: 0;
          font-size: inherit;
          color: inherit;
        }
        .copy {
          background-color: #17c964;
          display: inline-flex;
          justify-content: center;
          align-items: flex-start;
          width: calc(2 * 20px);
          color: inherit;
          transition: opacity 0.2s ease 0s;
          cursor: pointer;
          user-select: none;
        }
        .copy:hover {
          opacity: 0.7;
        }
      </style></div></div>"
`;

exports[`Snippet should work with multi-line 1`] = `
"<div class=\\"snippet \\"><pre>cd project</pre><pre>now</pre><div role=\\"button\\" tabindex=\\"-1\\" class=\\"tooltip \\"><span class=\\"copy\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"18\\" height=\\"18\\" viewBox=\\"0 0 24 24\\"><path fill=\\"#444\\" d=\\"M20 2H10c-1.103 0-2 .897-2 2v4H4c-1.103 0-2 .897-2 2v10c0 1.103.897 2 2 2h10c1.103 0 2-.897 2-2v-4h4c1.103 0 2-.897 2-2V4c0-1.103-.897-2-2-2zM4 20V10h10l.002 10H4zm16-6h-4v-4c0-1.103-.897-2-2-2h-4V4h10v10z\\"></path></svg></span><style>
        .tooltip {
          width: max-content;
          display: inherit;
        }
      </style></div><style>
        .snippet {
          display: flex;
          position: relative;
          width: initial;
          max-width: 100%;
          padding: calc(20px * 0.75) 20px;
          color: #000;
          background: #fff;
          border: 0px solid #eaeaea;
          border-radius: 14px;
        }
        pre {
          margin: 0;
          padding: 0;
          border: none;
          border-radius: 0;
          width: calc(100% - 2 * 20px);
          background-color: transparent;
          color: #000;
          font-size: 0.8125rem;
        }
        pre::before {
          content: '$ ';
          user-select: none;
        }
        pre :global(*) {
          margin: 0;
          padding: 0;
          font-size: inherit;
          color: inherit;
        }
        .copy {
          background-color: #fff;
          display: inline-flex;
          justify-content: center;
          align-items: flex-start;
          width: calc(2 * 20px);
          color: inherit;
          transition: opacity 0.2s ease 0s;
          cursor: pointer;
          user-select: none;
        }
        .copy:hover {
          opacity: 0.7;
        }
      </style></div>"
`;
